File: solutions/c/resistor-color-trio/1/resistor_color_trio.c
#include "resistor_color_trio.h"

resistor_value_t color_code(resistor_band_t colors[]){
    unsigned long long int num = 0;
    unsigned int n = colors[2], zeros = colors[2];
    resistor_value_t resistor;
    for (int i = 0; i < 2; i++){
        num *= 10;
        num += colors[i];
    }
    while (n){
        num *= 10;
        n--;
    }
    if (colors[1] == 0){
        if (colors[0] == 0) zeros = 0;
        else zeros++;
    }
    if (zeros < 3){
        resistor.value = num;
        resistor.unit = OHMS;
    }
    else if (zeros < 6){
        resistor.value = num/1000;
        resistor.unit = KILOOHMS;
    }
    else if (zeros < 9){
        resistor.value = num/1000000;
        resistor.unit = MEGAOHMS;
    }
    else{
        resistor.value = num/1000000000;
        resistor.unit = GIGAOHMS;
    }
    return resistor;   
}

File: solutions/julia/resistor-color-trio/1/resistor-color-trio.jl
const UNITS = [" ohms", " kiloohms", " megaohms", " gigaohms"]

const COLORS = Dict(
    "black"  => 0,
    "brown"  => 1,
    "red"    => 2,
    "orange" => 3,
    "yellow" => 4,
    "green"  => 5,
    "blue"   => 6,
    "violet" => 7,
    "grey"   => 8,
    "white"  => 9
)

function label(colors)
    col = Tuple(COLORS[color] for color in colors)
    sub_digits = (col[1] > 0 ? 1 : 0) + col[3]
    zeros = "0"^(col[3] % 3)
    if sub_digits % 3 == 0 && sub_digits > 0 && col[2] > 0
        number = "$(col[1]).$(col[2])"
    elseif sub_digits % 3 == 0
        number = "$(col[1] > 0 ? col[1] : col[2])"
    elseif col[1] > 0
        number = "$(col[1])$(col[2])$zeros"
    else
        number = "$(col[2])$zeros"
    end
    number * UNITS[1 + sub_digits ÷ 3]
end